### PR TITLE
feat: support specifying a subset of fields to read

### PR DESCRIPTION
**Pull request by matteo-pallini**
Being able to specify a subset of fields to read and only read these is a particularly useful feature. In particular when reading data stored in parquet files, where it is possible to read only the data needed, rather than all of all the time and then remove the one not needed.

## Key Changes

If a subset of fields is specified whenever the IO resources is initialised schema validation is going to be performed only on these fields

## Impact

* introduce a backward compatible feature

## Checklist

The following won't apply in all cases - mark `N/A` next to point if not needed.

- [ ] Module, function, class docstrings updated
- [ ] Interface documentation updated
- [ ] Semantic commits used for this PR, so that a CHANGELOG can be generated from them (don't delete your local branch! when tagging a new release from master)

## Release-Steps

- [ ] Merge PR
- [ ] Release new tag

### CODE REVIEW

File: dynamicio/core.py [hunk 1]
```diff
@@ -278,8 +278,13 @@ def _has_valid_dtypes(self, df: pd.DataFrame) -> bool:
...
         """
         dtypes = df.dtypes
+        cols_to_validate = [self.schema.columns[c] for c in self.options.get("columns", self.schema.column_names) if c in self.schema.column_names]
 
-        for col_info in self.schema.columns.values():
+        if len(cols_to_validate) < len(self.options.get("columns", [])):
+            logger.exception(f"Some of the columns passed {self.options['columns']} do not belong to the schema")
+            return False
+
+        for col_info in cols_to_validate:
```
Comment (Christos-Hadjinikolis) [inline dynamicio/core.py:287]: For me, it was a bit hard to parse this. I would rewrite as follows: 

```python
# Get the columns to validate. If no `"columns"` are specified in `self.options`, then default to the schema columns.
requested_columns = self.options.get("columns", self.schema.column_names)

cols_to_validate = []
missing_columns = False

# Iterate over the requested columns
for column in requested_columns:
    # If a requested column is in the schema, add it to the cols_to_validate list
    if column in self.schema.column_names:
        cols_to_validate.append(self.schema.columns[column])
    # If a requested column is not in the schema, log an exception and set missing_columns to True
    else:
        logger.exception(f"The column '{column}' does not belong to the schema")
        missing_columns = True

# If any columns were missing, return False
if missing_columns:
    return False
```
Thoughts?

**Review by Christos-Hadjinikolis — CHANGES_REQUESTED** [review]

**Comment by Christos-Hadjinikolis** [comment]
More importantly, have you tested this version of `dynamicio` with any repo that uses it?

File: dynamicio/core.py (hunk 1 above)
Comment (Christos-Hadjinikolis) [reply dynamicio/core.py:287]: If am not going too far with this, I would dare say you can even extract this logic into a: 

```python 
def __get_columns_to_validate(filtered_columns: List[str, ], default_columns=List[str]): 
   ...
```
method, for readability's sake.

File: tests/test_core.py [hunk 2]
```diff
@@ -776,21 +776,38 @@ def test_show_casting_warnings_flag_allows_casting_logs_to_be_printed_if_set_to_
...
+            assert "columns" in kwargs_values
+            assert kwargs_values["columns"] == ["id", "foo_name", "bar"]
+
+    @pytest.mark.unit
+    def test_columns_options_with_names_missing_from_schema_fails(self, s3_parquet_local_config):
+        with patch("dynamicio.mixins.with_local.pd.read_parquet"):
+            reader = ReadS3ParquetIO(source_config=s3_parquet_local_config, columns=["bar", "column_missing_from_schema"])
+            with pytest.raises(ColumnsDataTypeError):
+                reader.read()
+
```
[the remark this answers is not in the document]
Comment (matteo-pallini) [reply tests/test_core.py]: I am not a big fan of this.... but addressed as requested

**Review by Christos-Hadjinikolis — APPROVED** [review]
